rust/lance/src/dataset.rs

```diff
@@ -812,51 +815,84 @@ impl Dataset {
         Ok(counts.iter().sum())
     }
 
-    pub async fn take(&self, row_indices: &[usize], projection: &Schema) -> Result<RecordBatch> {
-        let mut sorted_indices: Vec<u32> =
-            Vec::from_iter(row_indices.iter().map(|indice| *indice as u32));
-        sorted_indices.sort();
+    pub async fn take(&self, row_indices: &[u64], projection: &Schema) -> Result<RecordBatch> {
+        let mut sorted_indices: Vec<usize> = (0..row_indices.len()).collect();
+        sorted_indices.sort_by_key(|&i| row_indices[i]);
 
-        let mut row_count = 0;
-        let mut start = 0;
-        let schema = Arc::new(ArrowSchema::from(projection));
-        let mut batches = Vec::with_capacity(sorted_indices.len());
-        for fragment in self.get_fragments().iter() {
-            if start >= sorted_indices.len() {
-                break;
-            }
+        let fragments = self.get_fragments();
 
-            let max_row_indices = row_count + fragment.count_rows().await? as u32;
-            if sorted_indices[start] < max_row_indices {
-                let mut end = start;
-                sorted_indices[end] -= row_count;
-                while end + 1 < sorted_indices.len() && sorted_indices[end + 1] < max_row_indices {
-                    end += 1;
-                    sorted_indices[end] -= row_count;
-                }
-                batches.push(
-                    fragment
-                        .take(&sorted_indices[start..end + 1], projection)
-                        .await?,
-                );
+        // We will split into sub-requests for each fragment.
+        let mut sub_requests: Vec<(&FileFragment, Range<usize>)> = Vec::new();
+        // We will remap the row indices to the original row indices, using a pair
+        // of (request position, position in request)
+        let mut remap_index: Vec<(usize, usize)> = vec![(0, 0); row_indices.len()];
+        let mut local_ids_buffer: Vec<u32> = Vec::with_capacity(row_indices.len());
 
-                // restore the row indices
-                for indice in sorted_indices[start..end + 1].iter_mut() {
-                    *indice += row_count;
+        let mut fragments_iter = fragments.iter();
+        let mut current_fragment = fragments_iter.next().ok_or_else(|| Error::InvalidInput {
+            source: "Called take on an empty dataset.".to_string().into(),
+        })?;
+        let mut current_fragment_len = current_fragment.count_rows().await?;
+        let mut curr_fragment_offset: u64 = 0;
+        let mut current_fragment_end = current_fragment_len as u64;
+        let mut start = 0;
+        let mut end = 0;
+
+        for index in sorted_indices {
+            // Get the index
+            let row_index = row_indices[index];
```

Comment:
Maybe `row_address` or `row_id`?

Reply:
These aren't `row_id`. They are more like `row_number`, in that there is no gap between fragments.